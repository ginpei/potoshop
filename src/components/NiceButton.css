.NiceButton {
}

.NiceButton {
  align-items: center;
  background-color: var(--color-main);
  border-radius: 3px;
  border: 1px solid var(--color-main);
  color: #fff;
  cursor: pointer;
  display: inline-flex;
  height: 50px;
  justify-content: center;
  min-width: 150px;
  text-decoration: none;
  vertical-align: middle;
}
.NiceButton:hover,
.NiceButton:active {
  background-color: var(--color-main);
  color: white;
}
.NiceButton:active {
  animation: poyon 300ms;
}

.NiceButton-icon {
  margin-right: 0.5em;
}
