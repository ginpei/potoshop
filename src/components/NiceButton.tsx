/**
 * @example
 * <NiceButton>Nice!</NiceButton>
 * @example
 * <NiceButton icon="thumbs-up">Nice!</NiceButton>
 * @example
 * <NiceButton icon="fa-thumbs-up">Nice!</NiceButton>
 * @example
 * <NiceButton onClick={onClick}>Nice!</NiceButton>
 */

import * as React from 'react';

interface INiceButtonProps {
  icon?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  primary?: boolean;
}
interface INiceButtonState {
}

class NiceButton extends React.Component<INiceButtonProps, INiceButtonState> {
  constructor (props: INiceButtonProps) {
    super(props);
    this.state = {
    };
    this.onClick = this.onClick.bind(this);
  }

  public render () {
    const className = [
      'NiceButton',
      'niceButtonBase',
      this.props.primary ? '-primary' : '',
    ].join(' ');

    const iconKey = this.props.icon;
    const iconClassName = iconKey && [
      'fa',
      iconKey.startsWith('fa-') ? iconKey : `fa-${iconKey}`,
      'niceButtonBase-leftIcon',
    ].join(' ');

    return (
      <button className={className} onClick={this.onClick}>
        {iconClassName && <i className={iconClassName} aria-hidden="true"/>}
        {this.props.children}
      </button>
    );
  }

  protected onClick (event: React.MouseEvent<HTMLButtonElement>) {
    this.props.onClick!(event);
  }
}

export default NiceButton;
