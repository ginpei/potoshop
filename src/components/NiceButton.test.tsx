import * as enzyme from 'enzyme';
import { shallow } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import NiceButton from './NiceButton';

// possible to move to test.js?
enzyme.configure({ adapter: new Adapter() });

describe('<NiceButton/>', () => {
  // type Wrapper = ShallowWrapper<PointerHandler['props'], PointerHandler['state'], PointerHandler>;

  it('renders with children', () => {
    const wrapper = shallow<NiceButton>(
      <NiceButton
        >Nice!</NiceButton>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders icon', () => {
    const wrapper = shallow<NiceButton>(
      <NiceButton
        icon="like"
        >Nice!</NiceButton>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.NiceButton-icon').prop('className'))
      .toBe('fa fa-like NiceButton-icon');
  });

  it('renders prefixed icon', () => {
    const wrapper = shallow<NiceButton>(
      <NiceButton
        icon="fa-like"
        >Nice!</NiceButton>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.NiceButton-icon').prop('className'))
      .toBe('fa fa-like NiceButton-icon');
  });

  it('invokes onClick callback', () => {
    const onClick = jest.fn();
    const wrapper = shallow<NiceButton>(
      <NiceButton
        onClick={onClick}
        >Nice!</NiceButton>,
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.simulate('click');
    expect(onClick).toBeCalled();
  });
});
