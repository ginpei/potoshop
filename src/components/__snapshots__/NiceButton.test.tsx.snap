// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NiceButton/> invokes onClick callback 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NiceButton
    onClick={[MockFunction]}
  >
    Nice!
  </NiceButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        "Nice!",
      ],
      "className": "NiceButton",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      undefined,
      "Nice!",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          "Nice!",
        ],
        "className": "NiceButton",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        undefined,
        "Nice!",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<NiceButton/> renders icon 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NiceButton
    icon="like"
  >
    Nice!
  </NiceButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <i
          aria-hidden="true"
          className="fa fa-like NiceButton-icon"
        />,
        "Nice!",
      ],
      "className": "NiceButton",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-like NiceButton-icon",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "Nice!",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <i
            aria-hidden="true"
            className="fa fa-like NiceButton-icon"
          />,
          "Nice!",
        ],
        "className": "NiceButton",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-hidden": "true",
            "className": "fa fa-like NiceButton-icon",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "Nice!",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<NiceButton/> renders prefixed icon 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NiceButton
    icon="fa-like"
  >
    Nice!
  </NiceButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <i
          aria-hidden="true"
          className="fa fa-like NiceButton-icon"
        />,
        "Nice!",
      ],
      "className": "NiceButton",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-like NiceButton-icon",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "Nice!",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <i
            aria-hidden="true"
            className="fa fa-like NiceButton-icon"
          />,
          "Nice!",
        ],
        "className": "NiceButton",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-hidden": "true",
            "className": "fa fa-like NiceButton-icon",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "Nice!",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<NiceButton/> renders with children 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NiceButton>
    Nice!
  </NiceButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        "Nice!",
      ],
      "className": "NiceButton",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      undefined,
      "Nice!",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          "Nice!",
        ],
        "className": "NiceButton",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        undefined,
        "Nice!",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
